### ISSUE TITLE
Celery autoscale not working

### ISSUE THREAD

**Issue by anishmac99** [celery/celery#4003]
I read from the celery documentation that we can autoscale concurrency by using autoscale option while starting a worker.

I have added 200 task in my queue and each task takes 5seconds to finish. I started worker with autoscale of 3 min concurrency and 15 max concurrency , but its starting with only 3 processes and I am not seeing any scale up in the number of processes.

Using celery with rabbitmq.

**Comment by anonymez** [comment]
We are facing the same issue with the same configuration.

We run celery with --autoscale=1024,4 but it is just stuck at 4.

 I read autoscaling has different issues and should be undocumented from 4.0. Am I right? Is there any suggest solution? Starting with an high concurrency is ok?
On the other hand I also read the autoscaling code is returning in issue https://github.com/celery/celery/issues/3412. Is there someone can point me to some clarification post or can explain me how autoscaling is going?

Thanks.

**Comment by bb-swapnil** [comment]
I am also getting the same issue as mentioned by @anishmac99

**Comment by ChillarAnand** [comment]
Any steps to reproduce it? I am not able to reproduce it. 

```
celery worker -l info -A t --autoscale=8,1
```
I have started a  process and when I queued tasks, it is scaling automatically.

**Comment by pruthvirajbb** [comment]
We have already queued more than 1000 messages in rabbitmq, Each of this process is just sleeping for 5 seconds and then we started celery using this command 
celery -A app  worker -l info  --autoscale=10,2

Though there are 1000 functions to be called, the number of workers are still 2 and doesn't scale up

**Comment by thedrow** [comment]
Autoscaling was removed in 4.x. Sorry.

**Comment by jacobwgillespie** [comment]
@thedrow are you talking about `--autoscale`?  It doesn't appear to be removed as far as I can tell:

https://github.com/celery/celery/blob/5af199c94f3bf08dfaa93c36c83c2147b5f15639/celery/bin/worker.py#L127-L134

https://github.com/celery/celery/blob/5af199c94f3bf08dfaa93c36c83c2147b5f15639/celery/worker/autoscale.py#L2-L10

**Comment by anonymez** [comment]
@jacobwgillespie, We are all here struggling with autoscale since documentation and help are wrong and not updated.

**Comment by thedrow** [comment]
The release notes for 4.0 says it is removed so it should be.
No one maintained it for a while and it has a lot of bugs which we don't have time to resolve.

**Comment by Summer-D** [comment]
If it's obsoleted, why autoscale still shows up in the 4.1 document?

**Comment by thedrow** [comment]
It should be removed.

**Comment by champ** [comment]
@thedrow I've scanned the changelogs for 4.0.x and 4.1.x - the only reference to `--autoscale` I found in those releases is this one in 4.0.x:

> Auto-scale didn’t always update keep-alive when scaling down.
> 
> Fix contributed by Philip Garnero.

In what exact version was this removed and in what changelog was this reported?

Obviously, the references in the documentation should be removed if it's not still an option.

**Comment by illagrenan** [comment]
I think this issue should be re-opened.

**A)** Autoscaling was really removed --> I wasn't able to find any info about this in changelogs. And I can see Autoscaling section in the latest doc: http://docs.celeryproject.org/en/latest/userguide/workers.html#autoscaling.
**B)** Autoscaling wasn't removed but is probably broken.

**Comment by thedrow** [comment]
PRs are welcome to fix autoscaling or to remove it if no fixes are possible.

**Comment by gsmethells** [comment]
If the documentation says it is there, then I do not see how anyone could argue that it isn't supported. It is a bug until the documentation reflects that it was deprecated (properly! where usage causes an error!) and the CHANGELOG calls it out in explicit terms. Fix this situation before you muddy the reputation of this project.

**Comment by patrickscottbest** [comment]
I am also a little mystified on the functionality of autoscaler.   Is it decided to remove based on the 2018 sprint candidate?  Also, if it's still available, i'd like a scale DOWN method!  If my threads are busy chewing on something, then why would i want to overburden the machine with additional workload...  can this be accomplished by reversing the numbering in the --autoscale parameter ?  Am I making any kind of sense here?

**Comment by ninja-saigon** [comment]
Thank @thedrow 
I have the same issue with auto-scaling until I come here. Are there any other options for us if we actually wanna scale celery worker? or may be we have to do it by ourselves.

**Comment by thedrow** [comment]
I'm super busy with 4.3 right now.
I don't have time to look into this before Celery 5.

PRs are welcome for fixes of specific problems though.

**Comment by gsmethells** [comment]
@thedrow What I recommended is not a huge time sink but will prevent the reputation of this project form being muddied. Simply remove the documentation for the autoscaling topic and add a CHANGELOG entry to the effect of why it was removed.

**Comment by dejlek** [comment]
Autoscaling is one of the coolest Celery features. I would rather remove all that useless support for various brokers and result backends than autoscaling...

**Comment by auvipy** [comment]
> Autoscaling is one of the coolest Celery features. I would rather remove all that useless support for various brokers and result backends than autoscaling...

agree. can you investigate some bug n reasons of them of autoscale?

**Comment by dejlek** [comment]
I can't promise, but I will try to spend some time looking at the code... Problem is - I do not know Celery internals enough so will probably need considerate amount of time in order to be able to understand the code...

**Comment by asfaltboy** [comment]
@thedrow could there have been some confusion with the removed auto-reload feature?

> The `--autoreload` feature has been removed.

I think that autoscaling should still work as described, as related tests are passing and no1 has quoted any errors.

To everyone here that has issues, can you confirm that you are using the default `Autoscaler` class? And if you are, how many reserved tasks do you allow per worker (a.k.a [the `worker_prefetch_multiplier` setting][1]).

I'm making an assumption that many people will follow the Optimisation guide above, and set the setting to 1 (I do that as well); this will limit the maximum possible scaling to the number of worker processes currently running. In order for scaling to grow beyond that, I assume prefetch should be at least 2, to allow scaling beyond current worker count.

Alternatively, an alternative sub-class of `Autoscaler` can be used to debug `_maybe_scale` and `qty` methods are called and that the expected values are correctly evaluated.

Ref: https://github.com/celery/celery/blob/955ddf9a000d510ccb3fe7cba9eb0bb39dedc28a/celery/worker/autoscale.py#L159-L163

[1]: http://docs.celeryproject.org/en/latest/userguide/optimizing.html#prefetch-limits

**Comment by auvipy** [comment]
can anyone help to improve the tests of auto scale

**Comment by diegopedroso1** [comment]
@anishmac99 did you scale your workers with the celery --autoscale?

**Comment by auvipy** [comment]
https://github.com/celery/celery/pull/6069 hey guys can you check this patch?

**Comment by whatserface** [comment]
So in the end is it deprecated or does it still work? In the first case, what should we use instead?

**Comment by hash-data** [comment]
@thedrow what is the conclusion at the end, can we use it or is it removed?

**Comment by thedrow** [comment]
The feature is not removed and we're waiting for someone to take a look at what's broken with it.
Contributions are welcome.

**Comment by rimu** [comment]
I just searched the codebase for mentions of 'autoscale' and there were very many. It has not been removed.

**Comment by dohrisalim** [comment]
I can investigate this autoscaling bug where workers don't scale up despite having 200 tasks queued. I understand the autoscale concurrency mechanism and will trace through the worker initialization and scaling logic to identify why the process count stays at minimum instead of increasing under load. This is marked as NEEDS HELP and represents a real production issue.

**Comment by auvipy** [comment]
please, Do and let us know
